hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java

```diff
@@ -1753,30 +1796,57 @@ private void reconcileChunksPerBlock(KeyValueContainer container, Pipeline pipel
           chunkByteBuffer.clear();
           chunkByteBuffer.limit(chunkLength);
           int bytesRead = blockInputStream.read(chunkByteBuffer);
+          // Make sure we read exactly the same amount of data we expected so it fits in the block.
           if (bytesRead != chunkLength) {
-            throw new IOException("Error while reading chunk data from block input stream. Expected length: " +
+            throw new IOException("Error while reading chunk data from peer " + peer + ". Expected length: " +
                 chunkLength + ", Actual length: " + bytesRead);
           }
 
           chunkByteBuffer.flip();
           ChunkBuffer chunkBuffer = ChunkBuffer.wrap(chunkByteBuffer);
+          ChunkInfo chunkInfo = ChunkInfo.getFromProtoBuf(chunkInfoProto);
+          chunkInfo.addMetadata(OzoneConsts.CHUNK_OVERWRITE, "true");
           writeChunkForClosedContainer(chunkInfo, blockID, chunkBuffer, container);
-          // In reconciling missing chunks which happens at the end of the block, we are expected to have holes in
-          // the blockData's chunk list because we continue to reconcile even if there are failures while reconciling
-          // chunks which is fine as we don't update the bcsId.
-          localChunksMap.put(chunkInfo.getOffset(), chunkInfoProto);
+          localOffset2Chunk.put(chunkOffset, chunkInfoProto);
+          if (LOG.isDebugEnabled()) {
+            LOG.debug("Successfully ingested chunk at offset {} into block {} of container {} from peer {}",
+                chunkOffset, localID, containerID, peer);
+          }
+          numSuccessfulChunks++;
         } catch (IOException ex) {
-          overwriteBcsId = false;
-          LOG.error("Error while reconciling chunk {} for block {} in container {}",
-              chunkOffset, blockID, containerData.getContainerID(), ex);
+          // The peer's chunk was expected to be healthy. Log a stack trace for more info as to why this failed.
+          LOG.error("Failed to ingest chunk at offset {} for block {} in container {} from peer {}",
+              chunkOffset, localID, containerID, peer, ex);
+          allChunksSuccessful = false;
         }
+        // Stop block repair once we fail to pull a chunk from the peer.
+        // Our write chunk API currently does not have a good way to handle writing around holes in a block.
+        if (!allChunksSuccessful) {
+          break;
+        }
+      }
+
+      // Do not update block metadata in this container if we did not ingest any chunks for the block.
+      if (!localOffset2Chunk.isEmpty()) {
+        BlockData putBlockData = BlockData.getFromProtoBuf(peerBlockData);
```

Comment:
If we already have a `localBlockData` we should use that instead of replacing it with `peerBlockData`. Ideally both should be the same, but it's better to update the existing one than to replace it with `peerBlockData`